src/segmented_buffer.rs

```diff
@@ -263,24 +379,47 @@ where
     }
 }
 
-impl<F> SegmentedPoolBuf<F, BytesMut, AllocBytesMutFn> {
-    pub fn into_bytes_stream(self) -> SegmentedBufBytes {
+impl<F> SegmentedPoolBuf<F, Buffer, AllocBufferFn> {
+    pub fn into_bytes_stream(mut self) -> SegmentedBufBytes {
+        self.buf
+            .bufs
+            .iter_mut()
+            .for_each(|b: &mut Reusable<Buffer>| {
+                let buf = std::mem::replace(b.get_mut(), Buffer::Write(BytesMut::new()));
+                let buf = buf.freeze();
+                let _ = std::mem::replace(b.get_mut(), buf);
+            });
         SegmentedBufBytes {
-            bufs: self
-                .buf
-                .bufs
-                .into_iter()
-                .map(move |mut b| Some(b.split().freeze()))
-                .collect(),
+            bufs: self.buf.bufs.into_iter().collect(),
         }
     }
 
     pub fn len(&self) -> usize {
         self.buf.len()
     }
+
+    fn duplicate(&self) -> Self {
+        let buf = SegmentedBuf::with_segment_size(self.buf.segment_size);
+        Self {
+            pool: self.pool.clone(),
+            buf,
+            buf_fut: None,
+            total_written: None,
+            pool_buf_max_size: self.pool_buf_max_size,
+        }
+    }
 }
 
-impl<F> Buf for SegmentedPoolBuf<F, BytesMut, AllocBytesMutFn> {
+impl<F> Clone for SegmentedPoolBuf<F, Buffer, AllocBufferFn> {
```

Comment:
From the API perspective, cloning/duplicating a pooled buffer doesn't make much sense. I think it should be something like "pool give me buffer" semantics.

Reply:
Here we're actually asking for a new buffer with the same contents as the old one, it's much like clone on a vector, with the pool taking the place of the vec's Alloc type param https://doc.rust-lang.org/src/alloc/vec.rs.html#301